src/libgit-utils.stanza

```diff
@@ -0,0 +1,197 @@
+defpackage slm/libgit-utils:
+  import core
+  import collections
+
+  import libgit2
+
+  import slm/utils
+  import slm/dependency
+  import slm/logging
+  import slm/git-dep
+
+;============================================================
+;====================== Libgit2 Utils =======================
+;============================================================
+
+public defstruct LibGitException <: Exception :
+  code: git_error_code
+  msg: ?
+
+public defmethod print (o:OutputStream, e:LibGitException):
+  print(o, "Libgit error: %_" % [msg(e)])
+
+; Finds and returns the object ID of the reference 'refish' in the repository at
+; 'path'
+; Throws an exception if unsuccessful
+public defn libgit-rev-parse (path:String, refish:String) -> String :
+  libgit2_init()
+
+  ; Open
+  val repo = match(libgit2_repository_open(path)):
+    (repo:GIT_REPOSITORY) :
+      repo
+    (e:git_error_code) :
+      val last-err-msg = message(libgit2_error_last())
+      libgit2_shutdown()
+      throw(LibGitException(e, "Failed to open git repository in %_: %_" % [path, last-err-msg]))
+
+  ; Rev-parse
+  val result = match(libgit2_revparse(repo, refish)):
+    (hash:String) : hash
+    (e:git_error_code) :
+      val last-err-msg = message(libgit2_error_last())
+      libgit2_repository_free(repo)
+      libgit2_shutdown()
+      throw(LibGitException(e, "Failed to parse reference %_: %_" % [refish, last-err-msg]))
+
+  ; Cleanup
+  libgit2_repository_free(repo)
+  libgit2_shutdown()
+
+  result
+
+; Clones the repository for the Git dependency and performs a forced checkout
+; of the tag of the dependency's version
+; Assumes the remote name is "origin"
+; Throws an exception if unsuccessful
+public defn libgit-fetch-dependency (dep:GitDependency) -> False :
+  val name = /name(dep)
+  val url = full-url-from-locator(locator(dep))
+  val path = /path(dep)
+  val tag = to-string("v%~" % [version(dep)])
+  val full-tag = to-string("refs/tags/%_" % [tag])
+  val refspec = to-string("+%_:%_" % [full-tag, full-tag])
+
+  ; Initialize libgit
+  libgit2_init()
+
+  ; Clone
+  val repo = match(libgit2_clone(url, path, 1)):
```

Comment:
I think `libgit2_*` wrapper functions should handle the error conversion to exception. You needing to handling each and every function call is not scalable. 

It also prevents you from being able to do: 

```
try:
   ... libgit2 calls here

catch():
   ... handling

finally: 
  libgit2_repository_free(repo)
  libgit2_shutdown()
```

which would be a much cleaner pattern.

Lets move this conversion to exception code into the `libgit2` library and clean up our interface code.

Reply:
In general I definitely prefer having cleaner interfaces, but in this case I think it would just add more hassle. I handle each `libgit` function's errors separately because they each require different reactions. For example, the `libgit2_repository_free` function should not be called if `libgit2_repository_open` or `libgit2_clone` themselves fail at the start. Some of the return codes are also "desirable" return values which I switch off of. For example, in my other branch, I use the return code of `libgit-revparse` to determine which style (v0.1.0 or 0.1.0) of tag exists in the local repository. I would rather not have to `try-catch` the result of that call to determine which tags exist. I also include information about the dependency and the requested user operation when throwing errors on the SLM side, information that would not be present on the `libgit2` library side unless I explicitly plumbed it over.

Reply:
> I would rather not have to try-catch the result of that call to determine which tags exist.

Can you expound on this ? What is the drawback here ?

Reply:
`libgit-revparse` returns the error code `GIT_ENOTFOUND` if a tag (or any reference-ish string in general) can not be resolved in the local repository. In my "non-v-prepended tags" branch (`JITX-8660/non-v-tag`), I check if the returned code is this in order to determine which tag style (`v0.1.0` or `0.1.0`) exists for a dependency. I would prefer to do this check with a `switch` or `if else` statement instead of a `try catch`.

Reply:
Actually the function is named `libgit2_revparse` not `libgit-revparse`, sorry for any confusion

Reply:
Right - but you haven't addressed my question:

```
 I would prefer to do this check with a switch or if else statement instead of a try catch.
```

Why ?

Reply:
It technically works, but I think it's better to reserve exceptions for when something has actually gone wrong. I think it is also more awkward to read a try-catch versus a switch. Also, I would still need a switch in the try-catch anyways to distinguish between a `GIT_ENOTFOUND` exception versus other libgit exceptions.